src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Headers.cs

```diff
@@ -287,4 +289,152 @@ public async Task SendAsync_WithZeroLengthHeaderName_Throws()
+                    {
+                        if (expectSuccess)
+                            throw;
+                    }
+
+                    if (!expectSuccess)
+                        return;
+
+                    Assert.All(requestData.Headers,
+                        h => Assert.False(h.HuffmanEncoded, "Expose raw decoded bytes once HuffmanEncoding is supported"));
+
+                    Encoding valueEncoding = Encoding.GetEncoding("ISO-8859-1");
+
+                    foreach ((string name, string[] values) in headers)
+                    {
+                        byte[] valueBytes = valueEncoding.GetBytes(string.Join(", ", values));
+
+                        Assert.Single(requestData.Headers,
+                            h => h.Name.Equals(name, StringComparison.OrdinalIgnoreCase) && h.Raw.AsSpan().IndexOf(valueBytes) != -1);
+                    }
+                })
+                .GetAwaiter().GetResult();
+
+            return RemoteExecutor.SuccessExitCode;
+        }
+
+        protected int SendAsync_ReceiveNonAsciiCharacters_Inner(string useAppCtxSwitchInner, string switchValueInner, string unicodeInner)
+        {
+            bool allowLatin1 = AllowLatin1Headers(useAppCtxSwitchInner, switchValueInner);
+            bool unicode = bool.Parse(unicodeInner);
+            bool expectSuccess = allowLatin1 && !unicode;
+
+            LoopbackServerFactory.CreateClientAndServerAsync(
+                async uri =>
+                {
+                    using var requestMessage = new HttpRequestMessage(HttpMethod.Get, uri) { Version = VersionFromUseHttp2 };
+                    using HttpClient client = CreateHttpClient();
+
+                    using HttpResponseMessage response = await client.SendAsync(requestMessage);
+
+                    if (!expectSuccess)
+                        return;
+
+                    Encoding valueEncoding = Encoding.GetEncoding("ISO-8859-1");
+                    foreach ((string name, string[] values) in GetNonAsciiTestHeaders(unicode))
+                    {
+                        IEnumerable<string> receivedValues = Assert.Single(response.Headers, h => h.Key.Equals(name, StringComparison.OrdinalIgnoreCase)).Value;
+                        string value = Assert.Single(receivedValues);
+
+                        string expected = valueEncoding.GetString(valueEncoding.GetBytes(string.Join(", ", values)));
+                        Assert.Equal(expected, value, StringComparer.OrdinalIgnoreCase);
+                    }
+                },
+                async server =>
+                {
+                    List<HttpHeaderData> headerData = GetNonAsciiTestHeaders(unicode)
+                        .Select(h => new HttpHeaderData(h.Name, string.Join(", ", h.Values), latin1: allowLatin1))
+                        .ToList();
+
+                    await server.HandleRequestAsync(headers: headerData);
+                })
+                .GetAwaiter().GetResult();
```

Comment:
`RemoteExecutor` can invoke an async method, we certainly do that in other places.